### PR TITLE
Support only-within-sequence attention for MosaicGPT

### PR DESCRIPTION
**Pull request by alextrott16**
Makes it possible to have attention restricted to tokens within the same source sequence when using pre-concatenated text dataloading.

- Adds `eos_token_id` and `bos_token_id` flags for the `text` dataloader, which instructs it to add `"sequence_id"` to the batch using one of those tokens as the separator.
- Adds `attn_uses_sequence_id` flag to MosaicGPT config/model and `sequence_id` to the forward arguments.
- Modifies the attention bias construction to restrict attention based on `sequence_id` if the model config is set to use that input.

125m comparison: https://wandb.ai/mosaic-ml/v004-125m
3b comparison: https://wandb.ai/mosaic-ml/v004-3b/table?workspace=user-alextrott (for profiling; not a full convergence run)

### CODE REVIEW

**Comment by vchiley** [comment]
cc @jfrankle

**Comment by alextrott16** [comment]
FYI, I think this PR has picked up some commits from the main branch that maybe weren't part of `release/v0.0.4` yet? There are some changed files that I didn't touch in any of my commits.

File: examples/common/text_data.py [hunk 1]
```diff
@@ -137,13 +137,65 @@ def __getitem__(self, idx: int) -> Dict[str, Any]:
...
+        return batch
+
+    @staticmethod
+    def sequence_id_from_bos(batch: Dict[str, torch.Tensor],
+                             split_token_id: int) -> Dict[str, torch.Tensor]:
+        is_bos = torch.eq(batch['input_ids'], split_token_id)
+        cumulative_bos = torch.cumsum(is_bos,
+                                      dim=1).to(batch['input_ids'].dtype)
+        batch['sequence_id'] = cumulative_bos
+        return batch
```
Comment (vchiley) [inline examples/common/text_data.py:189]: ❤️

File: examples/common/text_data.py [hunk 2]
```diff
@@ -137,13 +137,65 @@ def __getitem__(self, idx: int) -> Dict[str, Any]:
...
+            raise ValueError(
+                'Must supply a value for either eos_token_id or bos_token_id, but both were given.'
+            )
+        if eos_token_id is None:
+            self.split_token_id = bos_token_id
+            self.eos_mode = False
+        else:
+            self.split_token_id = eos_token_id
+            self.eos_mode = True
+        assert self.split_token_id is not None
```
Comment (vchiley) [inline examples/common/text_data.py:162]: informative value error instead of uninformative assert

File: examples/common/text_data.py [hunk 3]
```diff
@@ -137,13 +137,65 @@ def __getitem__(self, idx: int) -> Dict[str, Any]:
...
+        if (eos_token_id is not None) and (bos_token_id is not None):
+            raise ValueError(
+                'Must supply a value for either eos_token_id or bos_token_id, but both were given.'
+            )
+        if eos_token_id is None:
+            self.split_token_id = bos_token_id
+            self.eos_mode = False
+        else:
+            self.split_token_id = eos_token_id
+            self.eos_mode = True
```
Comment (vchiley) [inline examples/common/text_data.py:161]: Why use `self.split_token_id` and `self.eos_mode`?
Why not just set `self.eos_token_id` and `self.bos_token_id` then in call make the call based on those?

File: examples/common/text_data.py [hunk 4]
```diff
@@ -137,13 +137,65 @@ def __getitem__(self, idx: int) -> Dict[str, Any]:
...
+    def __call__(self, examples: List[Any]) -> Dict[str, torch.Tensor]:
+        batch = self.base_collator(examples)
+        if self.eos_mode:
+            return self.sequence_id_from_eos(batch, self.split_token_id)
+        else:
+            return self.sequence_id_from_bos(batch, self.split_token_id)
+
+    @staticmethod
+    def sequence_id_from_eos(batch: Dict[str, torch.Tensor],
+                             split_token_id: int) -> Dict[str, torch.Tensor]:
```
Comment (vchiley) [inline examples/common/text_data.py:173]: does this need to be a staticmethod?
if not, then you don't need to pass `split_token_id` you can just use `self.eos_token_id`
(assuming you use my previous suggestion)

same comment for `sequence_id_from_bos`

File: examples/common/text_data.py [hunk 5]
```diff
@@ -166,6 +218,19 @@ def build_text_dataloader(cfg: DictConfig, device_batch_size: int):
         mlm=mlm_probability is not None,
         mlm_probability=mlm_probability)
 
+    if cfg.dataset.get('eos_token_id') is not None:
+        if cfg.dataset.get('bos_token_id') is not None:
+            raise ValueError(
+                'Cannot use *both* EOS and BOS tokens for detecting sequence boundaries. ' +\
+                'Please supply `eos_token_id` if sequences end with an EOS token, or use ' +\
+                '`bos_token_id` if sequences start with a BOS token.'
+            )
```
Comment (vchiley) [inline examples/common/text_data.py:227]: isn't this already validated in `ConcatenatedSequenceCollatorWrapper`?

File: examples/llm/src/models/mosaic_gpt/mosaic_gpt.py [hunk 6]
```diff
@@ -153,8 +156,14 @@ def _attn_bias(self,
             assert isinstance(prefix_mask, torch.Tensor)  # pyright
             attn_bias = self._apply_prefix_mask(attn_bias, prefix_mask)
 
+        # If using torch or triton, we incorporate sequence_id (if appropriate)
+        if self.attn_uses_sequence_id and sequence_id is not None:
+            assert isinstance(attn_bias, torch.Tensor)  # pyright
+            # assert isinstance(sequence_id, torch.Tensor)  # pyright
```
Comment (vchiley) [inline examples/llm/src/models/mosaic_gpt/mosaic_gpt.py:162]:
```suggestion
```

File: examples/llm/src/models/mosaic_gpt/mosaic_gpt.py [hunk 7]
```diff
@@ -209,12 +218,34 @@ def _apply_prefix_mask(self, attn_bias: torch.Tensor,
...
+        # select seq_len subset of attn mask
+        attn_bias = attn_bias[..., :seq_len, :seq_len]
+
+        # Restrict attention to tokens that share the same value
+        # in sequence_id
+        cannot_attend = torch.logical_not(
+            torch.eq(sequence_id.view(-1, seq_len, 1),
+                     sequence_id.view(-1, 1, seq_len))).unsqueeze(1)
+        min_val = torch.finfo(attn_bias.dtype).min
+        attn_bias = attn_bias.masked_fill(cannot_attend, min_val)
```
Comment (vchiley) [inline examples/llm/src/models/mosaic_gpt/mosaic_gpt.py:238]: ❤️

File: examples/llm/src/models/mosaic_gpt/mosaic_gpt.py [hunk 8]
```diff
@@ -242,6 +273,19 @@ def forward(
                 'prefix_mask is a required argument when MosaicGPT is configured with prefix_lm=True.'
             )
 
+        if self.training:
```
Comment (vchiley) [inline examples/llm/src/models/mosaic_gpt/mosaic_gpt.py:276]: [Abhi mention that our eval set is also preconcat and should use this.](https://mosaicml.slack.com/archives/C05000N6VD3/p1680129645540909)
(Sorry I know I'm the one who lead you astray on this one)

File: examples/llm/src/models/mosaic_gpt/mosaic_gpt.py [hunk 9]
```diff
@@ -290,11 +334,13 @@ def forward(
             assert isinstance(self.transformer.emb_drop, nn.Module)  # pyright
             x = self.transformer.emb_drop(x_shrunk)
 
-        attn_bias, attention_mask = self._attn_bias(
-            device=x.device,
-            dtype=x.dtype,
-            attention_mask=attention_mask,
-            prefix_mask=prefix_mask)
+        with torch.no_grad():
```
Comment (vchiley) [inline examples/llm/src/models/mosaic_gpt/mosaic_gpt.py:337]: can we shove this into the `self._attn_bias` fn or better yet make it a decorator over that fn (see example [here](https://pytorch.org/docs/stable/generated/torch.no_grad.html) with decorator)

ex
```
    @torch.no_grad()
    def _attn_bias(self,
                   device,
                   dtype,
                   attention_mask: Optional[torch.ByteTensor] = None,
                   prefix_mask: Optional[torch.ByteTensor] = None):
                   prefix_mask: Optional[torch.ByteTensor] = None,
                   sequence_id: Optional[torch.LongTensor] = None):
```
when you define the fn above

File: examples/llm/src/models/mosaic_gpt/mosaic_gpt.py [hunk 10]
```diff
@@ -364,6 +410,11 @@ def prepare_inputs_for_generation(self,
             raise NotImplementedError(
                 'MosaicGPT does not support generation with right padding.')
 
+        if self.attn_uses_sequence_id and self.training:
```
Comment (vchiley) [inline examples/llm/src/models/mosaic_gpt/mosaic_gpt.py:413]: would we ever be using the generate fn if model is in `self.training` mode?

File: examples/llm/tests/test_model.py
Comment (vchiley) [inline examples/llm/tests/test_model.py:1]: i might be a good idea to add some prefixlm and seq_id tests here

**Review by vchiley — CHANGES_REQUESTED** [review]
There are a few things to address + a few nit

File: examples/llm/src/models/mosaic_gpt/mosaic_gpt.py (hunk 8 above)
Comment (alextrott16) [reply examples/llm/src/models/mosaic_gpt/mosaic_gpt.py:276]: I think we *have* to skip the input checking outside train mode. I'm fine with continuing to require `sequence_id` being supplied during training when `attn_uses_sequence_id==True`, but requiring it during eval mode is a huge burden.
Even though the input checking below this is skipped if the model is in eval mode, `sequence_id` will still be used if supplied (and the flag is turned on). That is, it'll handle our eval dataset correctly. So, I feel like this is a good compromise.

File: examples/llm/src/models/mosaic_gpt/mosaic_gpt.py (hunk 10 above)
Comment (alextrott16) [reply examples/llm/src/models/mosaic_gpt/mosaic_gpt.py:413]: We might not end up doing so, but it seems like a user could. HF generate does not put the model into eval mode, as far as I can tell.

File: examples/common/text_data.py (hunk 4 above)
Comment (alextrott16) [reply examples/common/text_data.py:173]: It doesn't need to be, but being a staticmethod makes testing a bit more convenient. I'll clean this part of the code up a bit and add some tests to actually take advantage of that choice, haha.

File: examples/llm/tests/test_model.py
Comment (alextrott16) [reply examples/llm/tests/test_model.py:1]: Added a combined prefix_lm/sequence_id test that should only pass if both of them are working as expected.

**Review by vchiley — APPROVED** [review]
lgtm

File: examples/llm/src/models/mosaic_gpt/mosaic_gpt.py (hunk 8 above)
Comment (vchiley) [reply examples/llm/src/models/mosaic_gpt/mosaic_gpt.py:276]: I guess if eval ds passes `sequence_id` to fwd, then it is used right?
